ava-labs/core-mobile
# CP-8375 Create push notifications for wallet activity

## Description

**Ticket: [CP-8375]** 

This PR enables push notifications for events on C-Chain when sending/receiving assets and approving transactions.
In essence it works like this:
- when user enters wallet we subscribe to `core-notification-sender-api` for balance changes with set of chain ids and addresses we want to be notified about
- the `core-notification-sender-api` service sends notifications via firebase cloud messaging (FCM) service which is then received by app and showed as native notification
- when user taps on notification Core mobile will open on C-chain portfolio page

There are some prerequisites for this to work:
- to use FCM we have opened firebase project and imported its SDK
- we also use firebase AppCheck service to protect `core-notification-sender-api` from abuse, meaning only our wallet can use this api
- user have to allow notification feature by accepting prompt which will show only once after this PR is merged, but can be turned off/on at any time through in-app settings

There were some issues too:
- AppCheck SDK breaks running tests by requiring some Native components so I added `tests/firebase/jestSetup.js` to avoid that
- there was existing implementation for notifications in DeeplinkContext which had to be a bit refactored to accept any kinds of data passing with `notifee` notifications

And there are some things left to do:
- push notification capability has to be enabled for iOS prod version at developer.apple.com
- registering with `core-notification-sender-api` has to be repeated if FCM token changes

 
## Screenshots/Videos


https://github.com/user-attachments/assets/0effb735-1937-485b-94b7-c3d802651e9d

https://github.com/user-attachments/assets/d030343d-3422-4b7d-8442-13afc2eafed8


## Testing
- check if permission popup shows on first run
- send some funds and observe if notification shows while app is running
- receive some funds and observe if notification shows while app is backgrounded
- do some approvals and observe same
- check if notifications can be turned off/on in settings

## Checklist

Please check all that apply (if applicable)


[CP-8375]: https://ava-labs.atlassian.net/browse/CP-8375?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
